File: go/128.go
package main

import (
	"math"
)

func longestConsecutive(nums []int) int {
	var numSet map[int]int = make(map[int]int)
	for _, v := range nums {
		numSet[v]++
	}
	var ret int = 0
	for _, v := range nums {
		if numSet[v-1] == 0 {
			var tmp int = v + 1
			for numSet[tmp] > 0 { // if still in nums
				tmp++
			}
			ret = int(math.Max(float64(tmp)-float64(v), float64(ret)))
		}
	}
	return ret
}

File: cpp/128.cpp
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
  int longestConsecutive(vector<int> &nums) {
    unordered_set<int> m(nums.begin(), nums.end());
    int ret = 0;
    for (int n : nums) {
      if (m.find(n - 1) == m.end()) {
        int tmp = n + 1;
        while (m.find(tmp) != m.end()) {
          tmp += 1;
        }
        if (tmp-n > ret)
          ret = tmp-n;
      }
    }
    return ret;
  }
};
